src/freenet/clients/http/ImageCreatorToadlet.java

```diff
@@ -60,51 +64,67 @@ public void handleMethodGET(URI uri, HTTPRequest req, ToadletContext ctx) throws
 			int requiredHeight = req.getParam("height").compareTo("") != 0 ? Integer.parseInt(req.getParam("height").endsWith("px")?req.getParam("height").substring(0, req.getParam("height").length()-2):req.getParam("height")) : DEFAULT_HEIGHT;
+			// Validate image size
+			if (requiredWidth <= 0 || requiredHeight <= 0) {
+				writeHTMLReply(ctx, 400, "Bad request", "Illegal argument");
+			}
+			if (requiredWidth > WIDTH_AND_HEIGHT_LIMIT || requiredHeight > WIDTH_AND_HEIGHT_LIMIT) {
+				writeHTMLReply(ctx, 400, "Bad request",
+						"Too large (max " + WIDTH_AND_HEIGHT_LIMIT + "x" + WIDTH_AND_HEIGHT_LIMIT + "px)");
+			}
 			// This is the image we are making
 			BufferedImage buffer = new BufferedImage(requiredWidth, requiredHeight, BufferedImage.TYPE_INT_RGB);
 			Graphics2D g2 = buffer.createGraphics();
 			g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
 			FontRenderContext fc = g2.getFontRenderContext();
-			// We then specify the maximum font size that fits in the image
-			// For this, we start at 1, and increase it, until it overflows. This-1 will be the font size
-			float size = 1;
-			g2.setFont(g2.getFont().deriveFont(size));
-			int width = 0;
-			int height = 0;
-			while (width < requiredWidth && height < requiredHeight) {
-				Rectangle2D bounds = g2.getFont().getStringBounds(text, fc);
-
-				// calculate the size of the text
-				width = (int) bounds.getWidth();
-				height = (int) bounds.getHeight();
-				g2.setFont(g2.getFont().deriveFont(++size));
-			}
-			g2.setFont(g2.getFont().deriveFont(size - 1));
+			specifyMaximumFontSizeThatFitsInImage(g2, fc, requiredWidth, requiredHeight, text);
 			Rectangle2D bounds = g2.getFont().getStringBounds(text, fc);
 			// actually do the drawing
 			g2.setColor(new Color(0, 0, 0));
-			g2.fillRect(0, 0, width, height);
+			g2.fillRect(0, 0, requiredWidth, requiredHeight);
 			g2.setColor(new Color(255, 255, 255));
 			// We position it to the center. Note that this is not the upper left corner
 			g2.drawString(text, (int) (requiredWidth / 2 - bounds.getWidth() / 2), (int) (requiredHeight / 2 + bounds.getHeight() / 4));
 
 			// Write the data, and send the modification data to let the client cache it
 			Bucket data = ctx.getBucketFactory().makeBucket(-1);
-			OutputStream os = data.getOutputStream();
-			try {
+			try (OutputStream os = data.getOutputStream()) {
 				ImageIO.write(buffer, "png", os);
-			} finally {
-				os.close();
 			}
-			MultiValueTable<String, String> headers=new MultiValueTable<String, String>();
+			MultiValueTable<String, String> headers = new MultiValueTable<>();
 			ctx.sendReplyHeadersStatic(200, "OK", headers, "image/png", data.size(), LAST_MODIFIED);
 			ctx.writeData(data);
 		}
 	}
 
 	@Override
 	public String path() {
-		return "/imagecreator/";
+		return ROOT_URL;
 	}
 
+	public void specifyMaximumFontSizeThatFitsInImage(Graphics2D g2, FontRenderContext fc,
```

Comment:
Okay, I think once you turn this into a package-private method (so you can reach it from the test) it should be okay to merge… I really don’t want to enlarge Fred’s API surface more than necessary. :)